File: packages/core/src/crypto.ts
/**
 * Client-side encryption primitives (SPEC.md §5.11).
 *
 * The byte-exact ciphertext envelope, the declared-type value serializer,
 * and AES-256-GCM encrypt/decrypt over WebCrypto (a global — this keeps
 * `@syncular/core` dependency-free). These are the codec-level pieces shared
 * by the wire boundary: the client encrypts a configured column's plaintext
 * value here before the row codec (§2.4) sees it as a `bytes` value, and
 * decrypts on apply. The row codec itself never touches this file.
 *
 * X25519 sealed-box key wrapping (the "async encryption" utilities) lives in
 * `@syncular/crypto` (it needs an audited curve implementation); this module
 * is the symmetric core and the envelope both sides agree on byte-for-byte.
 */
import { ByteReader, ByteWriter } from './bytes';

/** The pre-flip application type of an encrypted column (SPEC.md §5.11). */
export type DeclaredType =
  | 'string'
  | 'integer'
  | 'float'
  | 'boolean'
  | 'json'
  | 'blob_ref'
  | 'bytes';

/** A non-null row value, per the declared type. `crdt` is never encrypted. */
export type PlainValue = string | number | boolean | Uint8Array;

/** Envelope version byte (SPEC.md §5.11). */
export const ENVELOPE_VERSION = 0x01;

/** AES-GCM nonce length (96-bit, the GCM standard). */
export const NONCE_LENGTH = 12;

/** Symmetric key length: AES-256. */
export const KEY_LENGTH = 32;

/**
 * Client-local decrypt failure (SPEC.md §5.11, §10.3). Never on the wire —
 * the `client.` family is client-only. Raised at the apply seam for an
 * unknown envelope version, unknown `keyId`, GCM auth failure (wrong key), a
 * malformed envelope, or a post-decrypt value-parse failure. Not retryable.
 */
export class DecryptError extends Error {
  override readonly name = 'DecryptError';
  readonly code = 'client.decrypt_failed';
  readonly retryable = false;

  constructor(message: string) {
    super(message);
  }
}

/**
 * Injectable nonce source (SPEC.md §5.11 nonce discipline). Production uses
 * {@link secureRandomNonce}; crypto golden vectors inject a fixed nonce. A
 * fixed nonce MUST NOT be reachable from a production encode path.
 */
export type NonceSource = () => Uint8Array;

export function secureRandomNonce(): Uint8Array {
  const nonce = new Uint8Array(NONCE_LENGTH);
  crypto.getRandomValues(nonce);
  return nonce;
}

// ---------------------------------------------------------------------------
// Value serializer — declared type ⇄ canonical plaintext bytes (§5.11)
// ---------------------------------------------------------------------------

const textEncoder = new TextEncoder();
const textDecoder = new TextDecoder('utf-8', { fatal: true });

/**
 * Serialize a declared-type value to the canonical plaintext bytes fed to
 * GCM (SPEC.md §5.11 value serializer). Not a re-run of the row codec — this
 * is the self-describing per-`declaredType` encoding both cores agree on.
 */
export function serializePlain(
  declaredType: DeclaredType,
  value: PlainValue,
): Uint8Array {
  switch (declaredType) {
    case 'string':
    case 'json':
    case 'blob_ref':
      if (typeof value !== 'string') {
        throw new Error(`encrypted ${declaredType} column requires a string`);
      }
      return textEncoder.encode(value);
    case 'integer': {
      if (typeof value !== 'number' || !Number.isSafeInteger(value)) {
        throw new Error('encrypted integer column requires a safe integer');
      }
      const out = new Uint8Array(8);
      new DataView(out.buffer).setBigInt64(0, BigInt(value), true);
      return out;
    }
    case 'float': {
      if (typeof value !== 'number') {
        throw new Error('encrypted float column requires a number');
      }
      const out = new Uint8Array(8);
      new DataView(out.buffer).setFloat64(0, value, true);
      return out;
    }
    case 'boolean':
      if (typeof value !== 'boolean') {
        throw new Error('encrypted boolean column requires a boolean');
      }
      return new Uint8Array([value ? 1 : 0]);
    case 'bytes':
      if (!(value instanceof Uint8Array)) {
        throw new Error('encrypted bytes column requires a Uint8Array');
      }
      return value;
  }
}

/**
 * Parse the decrypted plaintext bytes back to a declared-type value (SPEC.md
 * §5.11). A `json`/`blob_ref` value is re-validated as the row codec would
 * (§2.4); any parse failure is a decrypt failure.
 */
export function deserializePlain(
  declaredType: DeclaredType,
  bytes: Uint8Array,
): PlainValue {
  switch (declaredType) {
    case 'string':
    case 'json':
    case 'blob_ref': {
      let text: string;
      try {
        text = textDecoder.decode(bytes);
      } catch {
        throw new DecryptError(
          `decrypted ${declaredType} value is not valid UTF-8`,
        );
      }
      if (declaredType === 'json') {
        try {
          JSON.parse(text);
        } catch {
          throw new DecryptError('decrypted json value does not parse');
        }
      }
      return text;
    }
    case 'integer': {
      if (bytes.length !== 8) {
        throw new DecryptError('decrypted integer must be 8 bytes');
      }
      const big = new DataView(bytes.buffer, bytes.byteOffset, 8).getBigInt64(
        0,
        true,
      );
      const num = Number(big);
      if (!Number.isSafeInteger(num)) {
        throw new DecryptError('decrypted integer outside safe range');
      }
      return num;
    }
    case 'float': {
      if (bytes.length !== 8) {
        throw new DecryptError('decrypted float must be 8 bytes');
      }
      return new DataView(bytes.buffer, bytes.byteOffset, 8).getFloat64(
        0,
        true,
      );
    }
    case 'boolean': {
      if (bytes.length !== 1 || (bytes[0] !== 0 && bytes[0] !== 1)) {
        throw new DecryptError('decrypted boolean must be one 0x00/0x01 byte');
      }
      return bytes[0] === 1;
    }
    case 'bytes':
      return bytes;
  }
}

// ---------------------------------------------------------------------------
// Envelope — 0x01 | keyIdLen(u8) | keyId | nonce(12) | ct+tag  (§5.11)
// ---------------------------------------------------------------------------

export interface Envelope {
  readonly keyId: string;
  readonly nonce: Uint8Array;
  /** AES-256-GCM ciphertext with the 16-byte tag appended. */
  readonly ciphertext: Uint8Array;
}

export function encodeEnvelope(envelope: Envelope): Uint8Array {
  const keyIdBytes = textEncoder.encode(envelope.keyId);
  if (keyIdBytes.length > 0xff) {
    throw new Error('keyId exceeds 255 UTF-8 bytes');
  }
  if (envelope.nonce.length !== NONCE_LENGTH) {
    throw new Error(`nonce must be ${NONCE_LENGTH} bytes`);
  }
  const writer = new ByteWriter();
  writer.u8(ENVELOPE_VERSION);
  writer.u8(keyIdBytes.length);
  writer.raw(keyIdBytes);
  writer.raw(envelope.nonce);
  writer.raw(envelope.ciphertext);
  return writer.finish();
}

export function decodeEnvelope(bytes: Uint8Array): Envelope {
  const reader = new ByteReader(bytes);
  let version: number;
  let keyIdLen: number;
  try {
    version = reader.u8();
    keyIdLen = reader.u8();
  } catch {
    throw new DecryptError('encrypted value: envelope truncated');
  }
  if (version !== ENVELOPE_VERSION) {
    throw new DecryptError(
      `encrypted value: unknown envelope version 0x${version.toString(16)}`,
    );
  }
  let keyIdBytes: Uint8Array;
  let nonce: Uint8Array;
  try {
    keyIdBytes = reader.raw(keyIdLen);
    nonce = reader.raw(NONCE_LENGTH);
  } catch {
    throw new DecryptError('encrypted value: envelope truncated');
  }
  // The remaining bytes are ciphertext+tag; must be at least the 16-byte tag.
  const ciphertext = bytes.subarray(2 + keyIdLen + NONCE_LENGTH);
  if (ciphertext.length < 16) {
    throw new DecryptError('encrypted value: ciphertext shorter than GCM tag');
  }
  let keyId: string;
  try {
    keyId = textDecoder.decode(keyIdBytes);
  } catch {
    throw new DecryptError('encrypted value: keyId is not valid UTF-8');
  }
  return { keyId, nonce, ciphertext: ciphertext.slice() };
}

// ---------------------------------------------------------------------------
// AES-256-GCM over WebCrypto
// ---------------------------------------------------------------------------

/**
 * Copy into a fresh `ArrayBuffer`-backed view. WebCrypto's `BufferSource`
 * type rejects a `Uint8Array<ArrayBufferLike>` (it could be SharedArrayBuffer);
 * a plain copy is unambiguously `ArrayBuffer`-backed.
 */
function toBuffer(bytes: Uint8Array): ArrayBuffer {
  const copy = new Uint8Array(bytes.length);
  copy.set(bytes);
  return copy.buffer;
}

async function importAesKey(key: Uint8Array): Promise<CryptoKey> {
  if (key.length !== KEY_LENGTH) {
    throw new Error(`AES-256-GCM key must be ${KEY_LENGTH} bytes`);
  }
  return crypto.subtle.importKey('raw', toBuffer(key), 'AES-GCM', false, [
    'encrypt',
    'decrypt',
  ]);
}

/**
 * Encrypt one declared-type value into a §5.11 envelope. `nonceSource`
 * defaults to a secure RNG; vectors inject a fixed nonce.
 */
export async function encryptValue(
  declaredType: DeclaredType,
  value: PlainValue,
  keyId: string,
  key: Uint8Array,
  nonceSource: NonceSource = secureRandomNonce,
): Promise<Uint8Array> {
  const plaintext = serializePlain(declaredType, value);
  const nonce = nonceSource();
  if (nonce.length !== NONCE_LENGTH) {
    throw new Error(`nonce source must yield ${NONCE_LENGTH} bytes`);
  }
  const cryptoKey = await importAesKey(key);
  const ct = new Uint8Array(
    await crypto.subtle.encrypt(
      { name: 'AES-GCM', iv: toBuffer(nonce), tagLength: 128 },
      cryptoKey,
      toBuffer(plaintext),
    ),
  );
  return encodeEnvelope({ keyId, nonce, ciphertext: ct });
}

/**
 * Decrypt a §5.11 envelope back to a declared-type value. `keyProvider`
 * resolves the envelope's `keyId` to key bytes; a missing key or GCM tag
 * mismatch is {@link DecryptError} (`client.decrypt_failed`).
 */
export async function decryptValue(
  declaredType: DeclaredType,
  envelopeBytes: Uint8Array,
  keyProvider: (keyId: string) => Uint8Array | undefined,
): Promise<PlainValue> {
  const envelope = decodeEnvelope(envelopeBytes);
  const key = keyProvider(envelope.keyId);
  if (key === undefined) {
    throw new DecryptError(
      `no key for keyId ${JSON.stringify(envelope.keyId)}`,
    );
  }
  const cryptoKey = await importAesKey(key);
  let plaintext: Uint8Array;
  try {
    plaintext = new Uint8Array(
      await crypto.subtle.decrypt(
        { name: 'AES-GCM', iv: toBuffer(envelope.nonce), tagLength: 128 },
        cryptoKey,
        toBuffer(envelope.ciphertext),
      ),
    );
  } catch {
    throw new DecryptError(
      `GCM authentication failed for keyId ${JSON.stringify(envelope.keyId)} (wrong key or corrupt ciphertext)`,
    );
  }
  return deserializePlain(declaredType, plaintext);
}

File: rust/crates/ssp2/src/crypto.rs
//! Client-side encryption primitives (SPEC.md §5.11) — the Rust half of the
//! cross-core §5.11 contract, byte-identical to `@syncular/core`'s `crypto.ts`.
//!
//! The ciphertext envelope, the declared-type value serializer, and
//! AES-256-GCM encrypt/decrypt. The row codec (`segment.rs`) never touches
//! this: an encrypted column has wire type `bytes` and the client's
//! encode/apply seam runs these functions on the plaintext value before/after
//! the codec. X25519 sealed-box key wrapping (the async-encryption utilities)
//! lives in `wrap.rs`.
//!
//! Gated behind the `e2ee` feature so lean builds skip the crypto crates.

use aes_gcm::aead::{Aead, KeyInit, Payload};
use aes_gcm::{Aes256Gcm, Nonce};

/// Envelope version byte (§5.11).
pub const ENVELOPE_VERSION: u8 = 0x01;
/// AES-GCM nonce length (96-bit).
pub const NONCE_LENGTH: usize = 12;
/// AES-256 key length.
pub const KEY_LENGTH: usize = 32;

/// The pre-flip declared type of an encrypted column (§5.11).
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum DeclaredType {
    String,
    Integer,
    Float,
    Boolean,
    Json,
    BlobRef,
    Bytes,
}

impl DeclaredType {
    pub fn from_name(name: &str) -> Option<Self> {
        match name {
            "string" => Some(DeclaredType::String),
            "integer" => Some(DeclaredType::Integer),
            "float" => Some(DeclaredType::Float),
            "boolean" => Some(DeclaredType::Boolean),
            "json" => Some(DeclaredType::Json),
            "blob_ref" => Some(DeclaredType::BlobRef),
            "bytes" => Some(DeclaredType::Bytes),
            _ => None,
        }
    }
}

/// A non-null declared-type value (mirrors ssp2 `ColumnValue`, minus crdt).
#[derive(Debug, Clone, PartialEq)]
pub enum PlainValue {
    String(String),
    Integer(i64),
    Float(f64),
    Boolean(bool),
    /// raw JSON document string (preserved verbatim, §2.4 tag 5)
    Json(String),
    /// raw canonical BlobRef string (§5.9.1)
    BlobRef(String),
    Bytes(Vec<u8>),
}

/// A §5.11 decrypt failure — `client.decrypt_failed`, never on the wire.
#[derive(Debug, Clone)]
pub struct DecryptError(pub String);

impl std::fmt::Display for DecryptError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "client.decrypt_failed: {}", self.0)
    }
}
impl std::error::Error for DecryptError {}

// -- value serializer: declared type ⇄ canonical plaintext bytes (§5.11) -----

/// Serialize a declared-type value to the canonical plaintext bytes fed to
/// GCM (§5.11 value serializer). Byte-identical to `serializePlain` in TS.
pub fn serialize_plain(value: &PlainValue) -> Result<Vec<u8>, String> {
    Ok(match value {
        PlainValue::String(s) | PlainValue::Json(s) | PlainValue::BlobRef(s) => {
            s.as_bytes().to_vec()
        }
        PlainValue::Integer(i) => i.to_le_bytes().to_vec(),
        PlainValue::Float(f) => f.to_le_bytes().to_vec(),
        PlainValue::Boolean(b) => vec![if *b { 1 } else { 0 }],
        PlainValue::Bytes(b) => b.clone(),
    })
}

/// Parse decrypted plaintext bytes back to a declared-type value (§5.11).
pub fn deserialize_plain(declared: DeclaredType, bytes: &[u8]) -> Result<PlainValue, DecryptError> {
    match declared {
        DeclaredType::String | DeclaredType::Json | DeclaredType::BlobRef => {
            let text = std::str::from_utf8(bytes)
                .map_err(|_| DecryptError("decrypted value is not valid UTF-8".to_owned()))?
                .to_owned();
            if declared == DeclaredType::Json {
                serde_json::from_str::<serde_json::Value>(&text)
                    .map_err(|_| DecryptError("decrypted json does not parse".to_owned()))?;
                Ok(PlainValue::Json(text))
            } else if declared == DeclaredType::BlobRef {
                Ok(PlainValue::BlobRef(text))
            } else {
                Ok(PlainValue::String(text))
            }
        }
        DeclaredType::Integer => {
            if bytes.len() != 8 {
                return Err(DecryptError("decrypted integer must be 8 bytes".to_owned()));
            }
            let mut a = [0u8; 8];
            a.copy_from_slice(bytes);
            Ok(PlainValue::Integer(i64::from_le_bytes(a)))
        }
        DeclaredType::Float => {
            if bytes.len() != 8 {
                return Err(DecryptError("decrypted float must be 8 bytes".to_owned()));
            }
            let mut a = [0u8; 8];
            a.copy_from_slice(bytes);
            Ok(PlainValue::Float(f64::from_le_bytes(a)))
        }
        DeclaredType::Boolean => {
            if bytes.len() != 1 || (bytes[0] != 0 && bytes[0] != 1) {
                return Err(DecryptError(
                    "decrypted boolean must be one 0x00/0x01 byte".to_owned(),
                ));
            }
            Ok(PlainValue::Boolean(bytes[0] == 1))
        }
        DeclaredType::Bytes => Ok(PlainValue::Bytes(bytes.to_vec())),
    }
}

// -- envelope: 0x01 | keyIdLen | keyId | nonce(12) | ct+tag  (§5.11) ---------

pub struct Envelope {
    pub key_id: String,
    pub nonce: [u8; NONCE_LENGTH],
    /// ciphertext with the 16-byte GCM tag appended.
    pub ciphertext: Vec<u8>,
}

pub fn encode_envelope(env: &Envelope) -> Result<Vec<u8>, String> {
    let key_id_bytes = env.key_id.as_bytes();
    if key_id_bytes.len() > 0xff {
        return Err("keyId exceeds 255 UTF-8 bytes".to_owned());
    }
    let mut out = Vec::with_capacity(2 + key_id_bytes.len() + NONCE_LENGTH + env.ciphertext.len());
    out.push(ENVELOPE_VERSION);
    out.push(key_id_bytes.len() as u8);
    out.extend_from_slice(key_id_bytes);
    out.extend_from_slice(&env.nonce);
    out.extend_from_slice(&env.ciphertext);
    Ok(out)
}

pub fn decode_envelope(bytes: &[u8]) -> Result<Envelope, DecryptError> {
    if bytes.len() < 2 {
        return Err(DecryptError("envelope truncated".to_owned()));
    }
    if bytes[0] != ENVELOPE_VERSION {
        return Err(DecryptError(format!(
            "unknown envelope version 0x{:02x}",
            bytes[0]
        )));
    }
    let key_id_len = bytes[1] as usize;
    let key_id_end = 2 + key_id_len;
    let nonce_end = key_id_end + NONCE_LENGTH;
    if bytes.len() < nonce_end + 16 {
        return Err(DecryptError("envelope truncated".to_owned()));
    }
    let key_id = std::str::from_utf8(&bytes[2..key_id_end])
        .map_err(|_| DecryptError("keyId is not valid UTF-8".to_owned()))?
        .to_owned();
    let mut nonce = [0u8; NONCE_LENGTH];
    nonce.copy_from_slice(&bytes[key_id_end..nonce_end]);
    Ok(Envelope {
        key_id,
        nonce,
        ciphertext: bytes[nonce_end..].to_vec(),
    })
}

// -- AES-256-GCM --------------------------------------------------------------

fn cipher(key: &[u8]) -> Result<Aes256Gcm, String> {
    if key.len() != KEY_LENGTH {
        return Err(format!("AES-256-GCM key must be {KEY_LENGTH} bytes"));
    }
    Aes256Gcm::new_from_slice(key).map_err(|e| format!("bad key: {e}"))
}

/// Encrypt one declared-type value into a §5.11 envelope with a caller-supplied
/// nonce (production supplies a CSPRNG nonce; vectors inject a fixed nonce).
pub fn encrypt_value(
    value: &PlainValue,
    key_id: &str,
    key: &[u8],
    nonce: [u8; NONCE_LENGTH],
) -> Result<Vec<u8>, String> {
    let plaintext = serialize_plain(value)?;
    let ct = cipher(key)?
        .encrypt(
            Nonce::from_slice(&nonce),
            Payload {
                msg: &plaintext,
                aad: &[],
            },
        )
        .map_err(|e| format!("aes-gcm encrypt failed: {e}"))?;
    encode_envelope(&Envelope {
        key_id: key_id.to_owned(),
        nonce,
        ciphertext: ct,
    })
}

/// Decrypt a §5.11 envelope back to a declared-type value. `key_provider`
/// resolves the envelope's `keyId`; a missing key or GCM tag mismatch is a
/// `DecryptError` (`client.decrypt_failed`).
pub fn decrypt_value(
    declared: DeclaredType,
    envelope_bytes: &[u8],
    key_provider: impl Fn(&str) -> Option<Vec<u8>>,
) -> Result<PlainValue, DecryptError> {
    let env = decode_envelope(envelope_bytes)?;
    let key = key_provider(&env.key_id)
        .ok_or_else(|| DecryptError(format!("no key for keyId {:?}", env.key_id)))?;
    let c = cipher(&key).map_err(DecryptError)?;
    let plaintext = c
        .decrypt(
            Nonce::from_slice(&env.nonce),
            Payload {
                msg: &env.ciphertext,
                aad: &[],
            },
        )
        .map_err(|_| {
            DecryptError(format!(
                "GCM authentication failed for keyId {:?} (wrong key or corrupt ciphertext)",
                env.key_id
            ))
        })?;
    deserialize_plain(declared, &plaintext)
}

#[cfg(test)]
mod tests {
    use super::*;

    const KEY: [u8; 32] = [7u8; 32];
    const NONCE: [u8; 12] = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11];

    fn provider(id: &str) -> Option<Vec<u8>> {
        if id == "k1" {
            Some(KEY.to_vec())
        } else {
            None
        }
    }

    #[test]
    fn round_trips_every_declared_type() {
        let cases = vec![
            (
                DeclaredType::String,
                PlainValue::String("hello 🔐".to_owned()),
            ),
            (DeclaredType::Json, PlainValue::Json("{\"a\":1}".to_owned())),
            (
                DeclaredType::BlobRef,
                PlainValue::BlobRef("{\"blobId\":\"x\"}".to_owned()),
            ),
            (DeclaredType::Integer, PlainValue::Integer(9007199254740991)),
            (
                DeclaredType::Integer,
                PlainValue::Integer(-9007199254740991),
            ),
            (DeclaredType::Float, PlainValue::Float(3.141592653589793)),
            (DeclaredType::Boolean, PlainValue::Boolean(true)),
            (DeclaredType::Bytes, PlainValue::Bytes(vec![0xde, 0xad])),
        ];
        for (ty, value) in cases {
            let env = encrypt_value(&value, "k1", &KEY, NONCE).unwrap();
            assert_eq!(env[0], ENVELOPE_VERSION);
            let back = decrypt_value(ty, &env, provider).unwrap();
            assert_eq!(back, value);
        }
    }

    #[test]
    fn wrong_key_fails() {
        let env = encrypt_value(&PlainValue::String("s".to_owned()), "k1", &KEY, NONCE).unwrap();
        let bad = decrypt_value(DeclaredType::String, &env, |id| {
            if id == "k1" {
                Some(vec![9u8; 32])
            } else {
                None
            }
        });
        assert!(bad.is_err());
    }

    #[test]
    fn fixed_nonce_is_deterministic() {
        let a = encrypt_value(&PlainValue::String("x".to_owned()), "k1", &KEY, NONCE).unwrap();
        let b = encrypt_value(&PlainValue::String("x".to_owned()), "k1", &KEY, NONCE).unwrap();
        assert_eq!(a, b);
    }
}
